Sources/Build/BuildPlan/BuildPlan.swift

```diff
@@ -904,6 +904,66 @@ extension BuildPlan {
 extension BuildPlan {
     fileprivate typealias Destination = BuildParameters.Destination
 
+    fileprivate enum TraversalNode: Hashable {
+        case package(ResolvedPackage)
+        case product(ResolvedProduct, BuildParameters.Destination)
+        case module(ResolvedModule, BuildParameters.Destination)
+
+        var destination: BuildParameters.Destination {
+            switch self {
+            case .package:
+                .target
+            case .product(_, let destination):
+                destination
+            case .module(_, let destination):
+                destination
+            }
+        }
+
+        static func `for`(
+            product: ResolvedProduct,
+            context destination: BuildParameters.Destination
+        ) -> Self {
+            switch product.type {
+            case .macro, .plugin:
+                .product(product, .host)
+            case .test:
+                .product(product, product.modules.contains(where: self.hasMacroDependency) ? .host : destination)
+            default:
+                .product(product, destination)
+            }
+        }
```

Comment:
nit: IMO initializers are more idiomatic for Swift
```suggestion
        init(
            product: ResolvedProduct,
            context destination: BuildParameters.Destination
        ) {
            switch product.type {
            case .macro, .plugin:
                self = .product(product, .host)
            case .test:
                self = .product(product, product.modules.contains(where: self.hasMacroDependency) ? .host : destination)
            default:
                self = .product(product, destination)
            }
        }
```